### PR TITLE
Ensure relocation target still tracked when start handoff

### PR DESCRIPTION
**Pull request by dnhatn**
If the master removes the relocating shard, but recovery isn't aware of it, then we can enter an invalid state where ReplicationTracker does not include the local shard.

```
java.lang.NullPointerException: null
	at org.elasticsearch.index.seqno.ReplicationTracker.getGlobalCheckpoint(ReplicationTracker.java:687) ~[elasticsearch-7.0.1.jar:7.0.1]
	at org.elasticsearch.index.seqno.ReplicationTracker.getAsLong(ReplicationTracker.java:692) ~[elasticsearch-7.0.1.jar:7.0.1]
	at org.elasticsearch.index.translog.TranslogWriter.syncNeeded(TranslogWriter.java:263) ~[elasticsearch-7.0.1.jar:7.0.1]
	at org.elasticsearch.index.translog.TranslogWriter.syncUpTo(TranslogWriter.java:341) ~[elasticsearch-7.0.1.jar:7.0.1]
	at org.elasticsearch.index.translog.TranslogWriter.sync(TranslogWriter.java:254) ~[elasticsearch-7.0.1.jar:7.0.1]
	at org.elasticsearch.index.translog.Translog.close(Translog.java:361) ~[elasticsearch-7.0.1.jar:7.0.1]
	at org.elasticsearch.core.internal.io.IOUtils.close(IOUtils.java:103) ~[elasticsearch-core-7.0.1.jar:7.0.1]
	at org.elasticsearch.core.internal.io.IOUtils.close(IOUtils.java:61) ~[elasticsearch-core-7.0.1.jar:7.0.1]
	at org.elasticsearch.index.engine.InternalEngine.closeNoLock(InternalEngine.java:2074) ~[elasticsearch-7.0.1.jar:7.0.1]
	at org.elasticsearch.index.engine.Engine.close(Engine.java:1732) ~[elasticsearch-7.0.1.jar:7.0.1]
	at org.elasticsearch.core.internal.io.IOUtils.close(IOUtils.java:103) ~[elasticsearch-core-7.0.1.jar:7.0.1]
	at org.elasticsearch.core.internal.io.IOUtils.close(IOUtils.java:61) ~[elasticsearch-core-7.0.1.jar:7.0.1]
	at org.elasticsearch.index.shard.IndexShard.close(IndexShard.java:1251) ~[elasticsearch-7.0.1.jar:7.0.1]
	at org.elasticsearch.index.IndexService.closeShard(IndexService.java:457) ~[elasticsearch-7.0.1.jar:7.0.1]
	at org.elasticsearch.index.IndexService.removeShard(IndexService.java:440) ~[elasticsearch-7.0.1.jar:7.0.1]
	at org.elasticsearch.indices.cluster.IndicesClusterStateService.failAndRemoveShard(IndicesClusterStateService.java:729) ~[elasticsearch-7.0.1.jar:7.0.1]
	at org.elasticsearch.indices.cluster.IndicesClusterStateService.handleRecoveryFailure(IndicesClusterStateService.java:721) ~[elasticsearch-7.0.1.jar:7.0.1]
	at org.elasticsearch.indices.cluster.IndicesClusterStateService.access$500(IndicesClusterStateService.java:105) ~[elasticsearch-7.0.1.jar:7.0.1]
```

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/seqno/ReplicationTracker.java [hunk 1]
```diff
@@ -1057,6 +1057,10 @@ public synchronized void completeRelocationHandoff() {
     public synchronized void activateWithPrimaryContext(PrimaryContext primaryContext) {
         assert invariant();
         assert primaryMode == false;
+        if (primaryContext.checkpoints.keySet().contains(shardAllocationId) == false) {
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/index/seqno/ReplicationTracker.java]: `primaryContext.checkpoints.containsKey()`?

**Review by ywelsch — CHANGES_REQUESTED** [review]
I wonder if we should instead pass the target allocation id to `startRelocationHandoff` which will then verify that the given allocation id exists in the returned set.

We can then assert in `activateWithPrimaryContext` that the `shardAllocationId` exists in the `PrimaryContext`.

Can you also add tests to ReplicationTrackerTests? TY

**Comment by dnhatn** [comment]
@ywelsch I've applied your suggestion. Can you take another look? Thank you!

File: server/src/main/java/org/elasticsearch/index/seqno/ReplicationTracker.java [hunk 2]
```diff
@@ -979,11 +980,15 @@ private synchronized void updateGlobalCheckpointOnPrimary() {
...
      */
-    public synchronized PrimaryContext startRelocationHandoff() {
+    public synchronized PrimaryContext startRelocationHandoff(String targetAllocationId) {
         assert invariant();
         assert primaryMode;
         assert handoffInProgress == false;
         assert pendingInSync.isEmpty() : "relocation handoff started while there are still shard copies pending in-sync: " + pendingInSync;
+        if (checkpoints.containsKey(targetAllocationId) == false) {
+            // can happen if the relocation target was removed from cluster but the recovery process isn't aware of that.
+            throw new IllegalStateException("relocation target [" + targetAllocationId + "] is no longer part of the group");
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/index/seqno/ReplicationTracker.java]: part of the *replication* group

File: server/src/main/java/org/elasticsearch/index/seqno/ReplicationTracker.java [hunk 3]
```diff
@@ -1037,6 +1042,12 @@ public synchronized void completeRelocationHandoff() {
     public synchronized void activateWithPrimaryContext(PrimaryContext primaryContext) {
         assert invariant();
         assert primaryMode == false;
+        // TODO: remove this check after backporting to 7.x
+        if (primaryContext.checkpoints.containsKey(shardAllocationId) == false) {
+            // can happen if the old primary was on an old version
+            assert indexSettings.getIndexVersionCreated().onOrAfter(Version.V_8_0_0);
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/index/seqno/ReplicationTracker.java]: should this be "before Version.V_8_0_0" and not "onOrAfter"?

File: server/src/main/java/org/elasticsearch/index/shard/IndexShard.java [hunk 4]
```diff
@@ -619,10 +619,11 @@ public IndexShardState markAsRecovering(String reason, RecoveryState recoverySta
      *
      * @param consumer a {@link Runnable} that is executed after operations are blocked
      * @throws IllegalIndexShardStateException if the shard is not relocating due to concurrent cancellation
+     * @throws IllegalStateException           if the relocation target is no longer part of the group
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/index/shard/IndexShard.java]: of the replication group

**Review by ywelsch — CHANGES_REQUESTED** [review]

File: server/src/main/java/org/elasticsearch/index/seqno/ReplicationTracker.java (hunk 3 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/index/seqno/ReplicationTracker.java]: sorry for this :(

**Comment by dnhatn** [comment]
@ywelsch I pushed changes.

**Review by ywelsch — APPROVED** [review]
LGTM

**Comment by dnhatn** [comment]
run elasticsearch-ci/oss-distro-docs

**Comment by dnhatn** [comment]
Thanks Yannick.

**Pull request merged**